Add UTXO and UTXOW rules for the ShelleyMA eras

nc6 opened the pull request:
This PR may be reviewed commit by commit.

It adds the overloaded rules for the new ShelleyMA (Mary and Allegra) eras. 

Where necessary, changes are made to the Shelley ledger to accommodate this:
- A few more functions are exported to allow defining `scriptsNeeded` for the new era.
- The UTXOW rule in Shelley is made era-specific to allow for overloading.

A question arises as to how to share the parts of the rule which do not change between eras. The approach I have taken in this PR is to try to follow the spec. So for the UTXO rule, which defined fully in the ShelleyMA spec (with new parts highlighted), I have re-implemented the whole rule in the shelley-ma package. For the UTXOW rule, which is explained in the spec as "The witnessing rule UTXOW does not itself change, but the function validateScript, which it
calls to evaluate the scripts prescribed by scriptsNeeded, does change, including a change to the
arguments passed to it, see", we modify the existing `utxoWitnessed` rule to take `scriptsNeeded` as a parameter.

nc6 committed:
Add UTXO rule for ShelleyMA eras.

nc6 committed:
Formatting

nc6 committed:
Parametrise utxowTransition over scriptsNeeded.

Also export some additional functions needed to implement
`scriptsNeeded` for the ShelleyMA eras.

nc6 committed:
Make UTXOW rule in Shelley era-specific.

This is needed, because we override this rule in ShelleyMA to use a
different `ScriptsNeeded`.

nc6 committed:
Add UTXOW rule for the ShelleyMA eras.

nc6 committed:
Script type for ShelleyMA

JaredCorduan commented:
it looks good to me so far! I see all the things I was expecting: checking the interval, no minting ada, checking forge scripts and Value scripts. I see that the `minUTxOValue` check is using the `scaledMinDeposit`, which is different from appendix B of the spec, but I do think I remember us saying that we'd not add those new pparams just yet.

nc6 commented on shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/Value.hs:
```diff
@@ -12,6 +12,7 @@ module Cardano.Ledger.ShelleyMA.Value
     Value (..),
     insert,
     lookup,
+    policies,
```
Note that this isn't actually the `Value` for `ShelleyMA`, it's just the `Value` for _Mary_. So it should be renamed as such

nc6 commented on shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/Rules/Utxow.hs:
```diff
@@ -0,0 +1,112 @@
+          scriptStakeCred
+          (filter requiresVKeyWitness certificates)
+      )
+    `Set.union` ((policyID `Set.map` (policies $ getField @"forge" txb)))
```
We will need a `policies` for `Coin` to get away with this.

nc6 committed:
Fix the type of 'inInterval'.

Also use camelCase.

nc6 committed:
Move 'Value' to 'Cardano.Ledger.Mary'.

The Allegra and Mary eras differ only in their `Value` type, with it
being set to Coin in Allegra and 'Value' in Mary. This commit reflects
that in the code structure, with Value now being specifically a Mary
type.

nc6 committed:
Make a few needed changes in TxBody.

- Specify that TxBody is the TxBody for the ShelleyMA eras.
- Derive a HashAnnotated instance.
- Fix up accessor names so that they match those from Shelley.

nc6 committed:
Be more explicit about kinds.

The aim here is to reduce the likelihood of failures due to kind
mismatch in the future, as well as to help with documentation.

nc6 committed:
Mary and Allegra top-level definitions.

This should allow these to be used by the consensus layer.

nc6 commented on shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/Rules/Utxo.hs:
```diff
@@ -0,0 +1,376 @@
+    <> getField @"forge" tx
+    <> (Val.inject $ refunds <> withdrawals)
+  where
+    -- balance (UTxO (Map.restrictKeys v (txins tx))) + refunds + withdrawals
```
Why is this comment still here?

nc6 commented on shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/Scripts.hs:
```diff
@@ -0,0 +1,81 @@
+    ScriptMSig s -> encodeListLen 2 <> toCBOR (0 :: Word8) <> toCBOR s
+    ScriptTimelock s -> encodeListLen 2 <> toCBOR (1 :: Word8) <> toCBOR s
+
+instance Era era => FromCBOR (Annotator (Script era)) where
```
Note that this serialisation is not correct. https://jira.iohk.io/browse/CAD-2145 covers the task of making the serialisation compatible, and @redxaxder has suggested a means by which this can be achieved.

nc6 replied:
(But in theory, we can go to a testnet without this)

nc6 committed:
Add test cradle to hie.yaml

nc6 committed:
Correct _poolPubKey to _poolId

This was somehow missed in #1945, and for some (unknown to me) reason,
CI didn't pick up on it. But when I try to build, it fails.

Regardless, this should be fixed!

uroboros reviewed (approved):
looking good!

mrBliss commented:
FYI, this looks good from the PoV of consensus. I have rebased #1944 locally on this PR and tried it with consensus (using the real eras, not just copies of Shelley!). Everything works, except for the serialisation generators that don't support a different `TxBody` or `Script` type yet, which means they can't be used for Allegra and Mary.

JaredCorduan reviewed (approved):
LGTM, thank you @nc6 !

polinavino commented on shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/Rules/Utxo.hs:
```diff
@@ -0,0 +1,376 @@
+          | out@(TxOut (AddrBootstrap addr) _) <- outputs,
+            bootstrapAddressAttrsSize addr > 64
+        ]
+  null outputsAttrsTooBig ?! OutputBootAddrAttrsTooBig outputsAttrsTooBig
```
do we need to check the size of the AssetID in the same way (it should have length <=32)
or is that getting enforced in the CBOR or CDDL?

nc6 replied:
I think Alex was going to ask you about this. Is it defined in the spec somewhere?

polinavino commented on shelley-ma/impl/src/Cardano/Ledger/Mary/Value.hs:
```diff
@@ -8,12 +8,13 @@
 {-# LANGUAGE TypeFamilies #-}
 
-module Cardano.Ledger.ShelleyMA.Value
+module Cardano.Ledger.Mary.Value
```
we are going with no special wrapper for Values that are allowed to have negative quantities? I am good with that, but am trying to figure out if that is explicitly the plan or we might add a wrapper later?

nc6 replied:
Good point. I've created https://jira.iohk.io/browse/CAD-2158 for this

polinavino replied:
thank you!

polinavino commented:
Looks good to me also!
